tests/unit/test_charm.py

```diff
@@ -1,11 +1,166 @@
+
+        # Cluster name is `cluster-<hash>`
+        self.assertTrue(peer_data["cluster-name"].isascii())
+        # Test passwords in content and length
+        required_passwords = ["root-password", "server-config-password", "cluster-admin-password"]
+        for password in required_passwords:
+            self.assertTrue(
+                peer_data[password].isalnum() and len(peer_data[password]) == PASSWORD_LENGTH
+            )
+
+        self.assertEqual(peer_data["configured"], "True")
+
+    @patch("charm.MySQLOperatorCharm._mysql", new_callable=PropertyMock)
+    def test_mysql_pebble_ready(self, _mysql_mock):
+        # Check if initial plan is empty
+        self.harness.set_can_connect("mysql", True)
+        initial_plan = self.harness.get_container_pebble_plan("mysql")
+        self.assertEqual(initial_plan.to_yaml(), "{}\n")
+
+        # Trigger pebble ready before leader election
+        self.harness.container_pebble_ready("mysql")
+        self.assertTrue(isinstance(self.charm.unit.status, WaitingStatus))
+
+        self.harness.set_leader()
+        self.charm._mysql = _mysql_mock
+        # Trigger pebble ready after leader election
+        self.harness.container_pebble_ready("mysql")
+        self.assertTrue(isinstance(self.charm.unit.status, ActiveStatus))
+
+        # After configuration run, plan should be populated
+        plan = self.harness.get_container_pebble_plan("mysql")
+        self.assertEqual(plan.to_dict()["services"], self.layer_dict["services"])
+
+    @patch("charm.MySQLOperatorCharm._mysql", new_callable=PropertyMock)
+    def test_mysql_pebble_ready_non_leader(self, _mysql_mock):
+        # Test pebble ready when not leader
+        # Expect unit to be in waiting status
+        self.harness.update_relation_data(self.peer_relation_id, "mysql/1", {"configured": "True"})
+
+        self.charm._mysql = _mysql_mock
+        self.harness.container_pebble_ready("mysql")
+        self.assertTrue(isinstance(self.charm.unit.status, WaitingStatus))
+
+    @patch("charm.MySQLOperatorCharm._mysql")
+    def test_mysql_pebble_ready_exception(self, _mysql_mock):
+        # Test exception raised in bootstrapping
+        self.harness.set_leader()
+        self.charm._mysql = _mysql_mock
+        _mysql_mock.bootstrap_instance.side_effect = MySQLBootstrapInstanceError
+        # Trigger pebble ready after leader election
+        self.harness.container_pebble_ready("mysql")
+
+        self.assertTrue(isinstance(self.charm.unit.status, BlockedStatus))
+
+    def test_on_config_changed(self):
+        # Test config changed set of cluster name
+        self.assertEqual(self.charm._peers.data[self.charm.app].get("cluster-name"), None)
+        self.harness.set_leader()
+        self.charm.on.config_changed.emit()
+        self.assertNotEqual(
+            self.charm._peers.data[self.charm.app].get("cluster-name"), "not_valid_cluster_name"
```

Comment:
I could not find this (maybe a) constant `not_valid_cluster_name` on the charm code.

Reply:
It's actually testing against the string `not_valid_cluster_name`. Generated cluster names will have the format `cluster-<10 chars hash>`, so the assert must pass when it differs. I've changed it on 8251bfe to ensure it fails when key is not present (by not using dict.get).